File: internal/plan/hydrate.go
package plan

import (
	"fmt"
	"sort"
	"strings"
)

const runtimeVersion = 1

type resolution struct {
	present bool
	value   string
}

func firstImageRef(entries []string) resolution {
	for _, e := range entries {
		if strings.TrimSpace(e) != "" {
			return resolution{true, e}
		}
	}
	return resolution{}
}

func asStringSlice(v interface{}) ([]string, bool) {
	switch t := v.(type) {
	case []string:
		return t, true
	case []interface{}:
		out := make([]string, 0, len(t))
		for _, x := range t {
			s, ok := x.(string)
			if !ok {
				return nil, false
			}
			out = append(out, s)
		}
		return out, true
	default:
		return nil, false
	}
}

func resolvePlaceholderValue(def Placeholder, raw interface{}) resolution {
	candidate := raw
	if candidate == nil {
		candidate = def.DefaultValue
	}
	if candidate == nil {
		return resolution{}
	}
	isImage := def.Type == "image" || def.Type == "image[]"
	if arr, ok := asStringSlice(candidate); ok {
		if len(arr) == 0 {
			return resolution{}
		}
		if isImage {
			return firstImageRef(arr)
		}
		return resolution{true, strings.Join(arr, "\n")}
	}
	s, ok := candidate.(string)
	if !ok {
		return resolution{}
	}
	if isImage {
		if strings.TrimSpace(s) == "" {
			return resolution{}
		}
		return resolution{true, s}
	}
	return resolution{true, s}
}

func sortElements(elements []CanvasElement) []CanvasElement {
	indexed := make([]CanvasElement, len(elements))
	copy(indexed, elements)
	sort.SliceStable(indexed, func(i, j int) bool {
		return indexed[i].ZIndex < indexed[j].ZIndex
	})
	return indexed
}

func strPtr(s string) *string { return &s }

func substituteTokens(content string, values map[string]interface{}) string {
	return inlineTokenRegex.ReplaceAllStringFunc(content, func(m string) string {
		sub := inlineTokenRegex.FindStringSubmatch(m)
		if len(sub) < 2 {
			return ""
		}
		token := sub[1]
		val, ok := values[token]
		if !ok || val == nil {
			return ""
		}
		if arr, ok := asStringSlice(val); ok {
			return strings.Join(arr, "\n")
		}
		if s, ok := val.(string); ok {
			return s
		}
		return fmt.Sprintf("%v", val)
	})
}

func hydrateArtifact(template Template, instanceID, layoutKey string, values map[string]interface{}, group *GroupRef) (ArtifactInstance, error) {
	if layoutKey == "" {
		layoutKey = "default"
	}
	layout, ok := template.Layouts[layoutKey]
	if !ok {
		return ArtifactInstance{}, fmt.Errorf("unknown layout %s on %s", layoutKey, template.ID)
	}
	defs := map[string]Placeholder{}
	for _, p := range template.Placeholders {
		defs[p.Key] = p
	}
	if values == nil {
		values = map[string]interface{}{}
	}
	effectiveValues := map[string]interface{}{}
	for k, v := range values {
		effectiveValues[k] = v
	}
	for _, p := range template.Placeholders {
		if _, ok := effectiveValues[p.Key]; !ok && p.DefaultValue != nil {
			effectiveValues[p.Key] = p.DefaultValue
		}
	}
	elements := make([]ResolvedElement, 0, len(layout.Elements))
	for _, element := range sortElements(layout.Elements) {
		style := element.Style
		if style == nil {
			style = map[string]interface{}{}
		}
		resolved := ResolvedElement{
			ID:     element.ID,
			Type:   element.Type,
			X:      element.X,
			Y:      element.Y,
			W:      element.W,
			H:      element.H,
			ZIndex: element.ZIndex,
			Style:  style,
		}
		if element.PlaceholderKey == nil || *element.PlaceholderKey == "" {
			if element.Type == "text" && element.Content != nil {
				tokens := ExtractInlineTokens(*element.Content)
				substituted := substituteTokens(*element.Content, effectiveValues)
				if len(tokens) > 0 {
					isSolelyToken := len(tokens) == 1 && strings.TrimSpace(*element.Content) == "{"+tokens[0]+"}"
					if isSolelyToken && strings.TrimSpace(substituted) == "" {
						if element.Required {
							return ArtifactInstance{}, fmt.Errorf("missing required placeholder %s", tokens[0])
						}
						continue
					}
					resolved.Text = &substituted
					if isSolelyToken {
						resolved.PlaceholderKey = strPtr(tokens[0])
					}
				} else {
					resolved.Text = element.Content
				}
			} else if (element.Type == "image" || element.Type == "image-placeholder") && element.ImageRef != nil {
				resolved.ImageURL = element.ImageRef
			}
			elements = append(elements, resolved)
			continue
		}
		def, ok := defs[*element.PlaceholderKey]
		if !ok {
			return ArtifactInstance{}, fmt.Errorf("undeclared placeholder %s", *element.PlaceholderKey)
		}
		value := resolvePlaceholderValue(def, effectiveValues[def.Key])
		if !value.present {
			if element.Required {
				return ArtifactInstance{}, fmt.Errorf("missing required placeholder %s", def.Key)
			}
			continue
		}
		resolved.PlaceholderKey = strPtr(def.Key)
		if element.Type == "text" {
			resolved.Text = strPtr(value.value)
		} else if element.Type == "image" || element.Type == "image-placeholder" {
			resolved.ImageURL = strPtr(value.value)
		}
		elements = append(elements, resolved)
	}
	inst := ArtifactInstance{
		RuntimeVersion: runtimeVersion,
		InstanceID:     instanceID,
		TemplateID:     template.ID,
		Label:          template.Label,
		BaseType:       template.BaseType,
		LayoutKey:      layoutKey,
		Layout: ResolvedLayout{
			AspectRatio:     layout.AspectRatio,
			BackgroundColor: layout.BackgroundColor,
			BackgroundImage: layout.BackgroundImage,
			Elements:        elements,
		},
		Group: group,
	}
	return inst, nil
}

func findResolvedText(instance ArtifactInstance, placeholderKey string) string {
	for _, el := range instance.Layout.Elements {
		if el.PlaceholderKey != nil && *el.PlaceholderKey == placeholderKey && el.Text != nil {
			return *el.Text
		}
	}
	return ""
}

func derivedLines(instance ArtifactInstance, title string) []string {
	type box struct {
		x, y float64
		text string
	}
	var boxes []box
	for _, el := range instance.Layout.Elements {
		if el.Type == "text" && el.Text != nil {
			boxes = append(boxes, box{el.X, el.Y, *el.Text})
		}
	}
	sort.SliceStable(boxes, func(i, j int) bool {
		if boxes[i].y != boxes[j].y {
			return boxes[i].y < boxes[j].y
		}
		return boxes[i].x < boxes[j].x
	})
	normalized := strings.TrimSpace(title)
	var lines []string
	for _, b := range boxes {
		for _, line := range strings.Split(b.text, "\n") {
			line = strings.TrimSpace(line)
			if line != "" && line != normalized {
				lines = append(lines, line)
			}
		}
	}
	return lines
}

File: src/lib/artifacts/hydrate.ts
/**
 * The single placeholder-resolution point.
 *
 * Turns an authored `ArtifactTemplate` plus weekly values into a fully resolved
 * `ArtifactInstance`. Nothing downstream of this module resolves placeholders,
 * applies defaults or reorders elements.
 */
import type {
  ArtifactLayout,
  ArtifactTemplate,
  CanvasElement,
  PlaceholderDefinition,
} from '@/lib/registry/types';
import {
  ARTIFACT_RUNTIME_VERSION,
  ArtifactHydrationError,
  type ArtifactGroupRef,
  type ArtifactInstance,
  type ArtifactLayoutKey,
  type ResolvedElement,
  type ResolvedLayout,
  type ResolvedStyle,
} from './runtime-contract';
import { requireTemplate, type RegistrySnapshot } from './registry-snapshot';
import { extractInlineTokens, INLINE_TOKEN_REGEX } from '@/lib/registry/placeholder-catalog';

export type PlaceholderValue = string | string[] | null | undefined;

export type PlaceholderValues = Readonly<Record<string, PlaceholderValue>>;

export type ArtifactHydrationRequest = {
  instanceId: string;
  templateId: string;
  layoutKey?: ArtifactLayoutKey;
  values?: PlaceholderValues;
  group?: ArtifactGroupRef;
};

export type HydrateOptions = {
  instanceId: string;
  layoutKey?: ArtifactLayoutKey;
  values?: PlaceholderValues;
  group?: ArtifactGroupRef;
};

type Resolution = { present: true; value: string } | { present: false };

const ABSENT: Resolution = { present: false };

function firstImageRef(entries: readonly string[]): Resolution {
  for (const entry of entries) {
    if (typeof entry === 'string' && entry.trim()) return { present: true, value: entry };
  }
  return ABSENT;
}

/**
 * `text` → string, `text[]` → newline join, `image` / `image[]` → first URL.
 * Empty string counts as a supplied text value; empty array / blank URL do not.
 */
function resolvePlaceholderValue(
  definition: PlaceholderDefinition,
  raw: PlaceholderValue
): Resolution {
  const candidate = raw ?? definition.defaultValue;
  if (candidate === undefined || candidate === null) return ABSENT;

  const isImage = definition.type === 'image' || definition.type === 'image[]';

  if (Array.isArray(candidate)) {
    if (candidate.length === 0) return ABSENT;
    if (isImage) return firstImageRef(candidate);
    return { present: true, value: candidate.join('\n') };
  }

  if (isImage) {
    return candidate.trim() ? { present: true, value: candidate } : ABSENT;
  }

  return { present: true, value: candidate };
}

function sortElements(elements: readonly CanvasElement[]): CanvasElement[] {
  return elements
    .map((element, index) => ({ element, index }))
    .sort((a, b) => a.element.zIndex - b.element.zIndex || a.index - b.index)
    .map((entry) => entry.element);
}

function toResolvedStyle(element: CanvasElement): ResolvedStyle {
  return { ...(element.style ?? {}) };
}

function baseResolvedElement(element: CanvasElement): ResolvedElement {
  return {
    id: element.id,
    type: element.type,
    x: element.x,
    y: element.y,
    w: element.w,
    h: element.h,
    zIndex: element.zIndex,
    style: toResolvedStyle(element),
  };
}

function substituteTokens(content: string, values: PlaceholderValues): string {
  return content.replace(INLINE_TOKEN_REGEX, (_match, token) => {
    const v = values[token];
    if (v === undefined || v === null) return '';
    if (Array.isArray(v)) return v.join('\n');
    return String(v);
  });
}

function resolveLayout(
  template: ArtifactTemplate,
  layout: ArtifactLayout,
  layoutKey: ArtifactLayoutKey,
  instanceId: string,
  values: PlaceholderValues
): ResolvedLayout {
  const definitions = new Map(template.placeholders.map((p) => [p.key, p]));
  const effectiveValues: Record<string, PlaceholderValue> = { ...values };
  for (const p of template.placeholders) {
    if (effectiveValues[p.key] === undefined && p.defaultValue !== undefined) {
      effectiveValues[p.key] = p.defaultValue;
    }
  }

  const elements: ResolvedElement[] = [];

  for (const element of sortElements(layout.elements)) {
    const resolved = baseResolvedElement(element);

    if (!element.placeholderKey) {
      if (element.type === 'text') {
        if (typeof element.content === 'string') {
          const tokens = extractInlineTokens(element.content);
          const substituted = substituteTokens(element.content, effectiveValues);
          if (tokens.length > 0) {
            // Check if the whole content was just token(s) and rendered completely empty
            const isSolelyToken = tokens.length === 1 && element.content.trim() === `{${tokens[0]}}`;
            if (isSolelyToken && substituted.trim() === '') {
              if (element.required) {
                throw new ArtifactHydrationError('Missing required placeholder value', {
                  instanceId,
                  templateId: template.id,
                  layoutKey,
                  elementId: element.id,
                  placeholderKey: tokens[0],
                });
              }
              // Optional token rendered empty -> omit element
              continue;
            }
            resolved.text = substituted;
            if (isSolelyToken) {
              resolved.placeholderKey = tokens[0];
            }
          } else {
            resolved.text = element.content;
          }
        }
      } else if (element.type === 'image' || element.type === 'image-placeholder') {
        if (typeof element.imageRef === 'string') resolved.imageUrl = element.imageRef;
      }
      elements.push(resolved);
      continue;
    }

    const definition = definitions.get(element.placeholderKey);
    if (!definition) {
      throw new ArtifactHydrationError('Element references an undeclared placeholder', {
        instanceId,
        templateId: template.id,
        layoutKey,
        elementId: element.id,
        placeholderKey: element.placeholderKey,
      });
    }

    const value = resolvePlaceholderValue(definition, effectiveValues[definition.key]);
    if (!value.present) {
      // The layout — not the placeholder declaration — decides whether a slot may
      // stay empty: a required element must have content, everything else is
      // simply dropped so optional weekly data (photos, lyric labels, sermon
      // titles) can be absent without failing the whole plan.
      if (element.required) {
        throw new ArtifactHydrationError('Missing required placeholder value', {
          instanceId,
          templateId: template.id,
          layoutKey,
          elementId: element.id,
          placeholderKey: definition.key,
        });
      }
      continue;
    }

    resolved.placeholderKey = definition.key;
    if (element.type === 'text') {
      resolved.text = value.value;
    } else if (element.type === 'image' || element.type === 'image-placeholder') {
      resolved.imageUrl = value.value;
    }
    elements.push(resolved);
  }

  return {
    aspectRatio: layout.aspectRatio,
    backgroundColor: layout.backgroundColor,
    ...(layout.backgroundImage ? { backgroundImage: layout.backgroundImage } : {}),
    elements,
  };
}

export function hydrateArtifact(
  template: ArtifactTemplate,
  options: HydrateOptions
): ArtifactInstance {
  const layoutKey: ArtifactLayoutKey = options.layoutKey ?? 'default';
  const layout = template.layouts[layoutKey];
  if (!layout) {
    throw new ArtifactHydrationError('Unknown layout key for template', {
      instanceId: options.instanceId,
      templateId: template.id,
      layoutKey,
    });
  }

  return {
    runtimeVersion: ARTIFACT_RUNTIME_VERSION,
    instanceId: options.instanceId,
    templateId: template.id,
    label: template.label,
    baseType: template.baseType,
    layoutKey,
    layout: resolveLayout(
      template,
      layout,
      layoutKey,
      options.instanceId,
      options.values ?? {}
    ),
    ...(options.group ? { group: options.group } : {}),
  };
}

export function hydrateArtifactFromSnapshot(
  snapshot: RegistrySnapshot,
  request: ArtifactHydrationRequest
): ArtifactInstance {
  const template = requireTemplate(snapshot, request.templateId, request.instanceId);
  return hydrateArtifact(template, {
    instanceId: request.instanceId,
    layoutKey: request.layoutKey,
    values: request.values,
    group: request.group,
  });
}
